core/internal/tensorboard/tfeventconverter.go

```diff
@@ -0,0 +1,144 @@
+package tensorboard
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/wandb/wandb/core/internal/tensorboard/tbproto"
+	"github.com/wandb/wandb/core/pkg/observability"
+	"github.com/wandb/wandb/core/pkg/service"
+)
+
+// TFEventConverter converts TF events into W&B history records.
+type TFEventConverter struct {
+	// Namespace is a prefix to add to all events.
+	Namespace string
+
+	// pluginNameByTag tracks the plugin name for each summary value tag.
+	//
+	// tfevents files may only contain a `metadata` field on the first
+	// occurrence of each tag to save space.
+	//
+	// See also the comment on the `metadata` field of the `Summary` proto.
+	pluginNameByTag map[string]string
+}
+
+// tagAndJSON is a tag and a JSON representation of a value.
+type tagAndJSON struct {
+	tag  string
+	json string
+}
+
+// Convert returns a W&B history record corresponding to a TF event.
+//
+// Returns nil if there's no relevant history data in the event.
+// Errors are logged via the logger and the corresponding data is ignored.
+func (h *TFEventConverter) Convert(
+	event *tbproto.TFEvent,
+	logger *observability.CoreLogger,
+) *service.HistoryRecord {
+	// Maps slash-separated tags to JSON values.
+	jsonData := make([]tagAndJSON, 0, len(event.GetSummary().GetValue()))
+
+	for _, value := range event.GetSummary().GetValue() {
+		tag := h.withNamespace(value.GetTag())
+
+		if h.rememberPluginName(tag, value) == "scalars" {
+			jsonData = processScalars(jsonData, tag, value, logger)
+		}
+	}
+
+	if len(jsonData) == 0 {
+		return nil
+	}
+
+	return h.toHistoryRecord(
+		jsonData,
+		event.Step,
+		event.WallTime,
+	)
+}
+
+// rememberPluginName returns the plugin name associated to the value.
+//
+// This returns the name stored in the value, or else the name stored most
+// recently for the tag.
+func (h *TFEventConverter) rememberPluginName(tag string, value *tbproto.Summary_Value) string {
+	if h.pluginNameByTag == nil {
+		h.pluginNameByTag = make(map[string]string)
+	}
+
+	if name := value.GetMetadata().GetPluginData().GetPluginName(); name != "" {
+		h.pluginNameByTag[tag] = name
+		return name
+	}
+
+	return h.pluginNameByTag[tag]
```

Comment:
do you need to check that `tag` is not an empty string?

Reply:
`tag` must be an actual tag, so not an empty string, for this function to be called

We also don't need to check that the `tag` is contained in the map---if not, then we return an empty plugin name which gets ignored. So it works even if the `tag` were an empty string.

Reply:
got it! thanks!